### PR TITLE
Extend events API to support child devices

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

This PR captures the code changes for extending the `events` APIs to child devices.
This includes:

-  If the `child device` is already registered with the `Cumulocity` cloud, then the `event message` to a `child device` will be sent to the `Cumulocity` cloud.
-   If  its not registered, then the `device will be registered first` and then the `event message` will be forwarded to the `Cumulocity` cloud.
- If the `child device` is not registered with the `Cumulocity` cloud and if the `event message` size is bigger than the supported `MQTT message size` then an error will be returned asking for "registering the child device".

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>https://github.com/thin-edge/thin-edge.io/issues/1197

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (rina23q) [inline crates/core/tedge_mapper/src/c8y/converter.rs:232]: I think you don't need to specify the type of `child_id`.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (rina23q) [inline crates/core/tedge_mapper/src/c8y/converter.rs:301]: Move this `Ok(vec)` to after line 279. Then you can remove all `return` from the `else` block.

```rust
        if input.payload_bytes().len() < self.size_threshold.0 {
            vec.push(message)
            Ok(vec)
       } else {
       // remove all `return` from this block
       }
```

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (rina23q) [inline crates/core/tedge_mapper/src/c8y/converter.rs:248]: Is this comment applicable for child device case? From your code, it doesn't look like that.

File: crates/core/thin_edge_json/src/event.rs
Comment (rina23q) [inline crates/core/thin_edge_json/src/event.rs:58]: It must be "3" or "4".

**Review by rina23q — COMMENTED** [review]
There is a room of improvement.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (matthiasbeyer) [inline crates/core/tedge_mapper/src/c8y/converter.rs:229]:
```suggestion
        let mut vec = Vec::new();
```

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (matthiasbeyer) [inline crates/core/tedge_mapper/src/c8y/converter.rs:232]:
```suggestion
        let child_id = get_child_id_from_event_topic(&input.topic.name)?;
```

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (matthiasbeyer) [inline crates/core/tedge_mapper/src/c8y/converter.rs:898]:
```suggestion
        None => Ok(None),
```

Just a small nitpick here: I'd rather be explicit here. This reduces the mental load when reading the code, as one can clearly see what is returned without having to look at the top of the `match` block again.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (matthiasbeyer) [inline crates/core/tedge_mapper/src/c8y/converter.rs:297]:
```suggestion
            if let Some(id) = child_id {
                if self.children.contains(&id) {
                    let _ = self.http_proxy.send_event(c8y_event).await?;
                    Ok(vec![])
                } else {
                    Err(ConversionError::ChildDeviceNotRegistered { id: id.to_string() })
                }
            } else {
                // Parent device
                let _ = self.http_proxy.send_event(c8y_event).await?;
                Ok(vec![])
            }
```

This is just a suggestion, your code looks absolutely fine! But as I've seen, you're using `match` a lot even if you're handling `Option<_>` - maybe you just didn't know that you can also do a `if let Some(thing) = optional_thing {}`!

As said in the introductory message: Feel free to ignore me entirely :wink:

**Review by matthiasbeyer — COMMENTED** [review]
Hi!

I have found some _very_ minor things, feel free to completely ignore my comments! :+1:

All of the below are just nitpicks and are not relevant in the grand scheme, I guess!

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: Refactored the code.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: refactored the code and added tests.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: Fixed

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs:248]: Yes, that's right. It's not applicable here.

File: crates/core/thin_edge_json/src/event.rs
Comment (PradeepKiruvale) [reply crates/core/thin_edge_json/src/event.rs]: Fixed

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: Refactored the code.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: Thanks for your comments on this PR. This will really help me to learn more (idiomatic) things about Rust.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: Yeah, fixed it.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: Updated the code.

File: crates/core/thin_edge_json/src/event.rs
Comment (didier-wenzek) [inline crates/core/thin_edge_json/src/event.rs:56]: Adding this `may_be_child` parameter is really a bad idea. This child id is extracted from the second argument i.e the topic name.

This makes the code more complex with:
* two places where the topic name is decoded,
* possible inconsistency when the two child device ids (given by the first and third arguments) are different.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (didier-wenzek) [reply crates/core/tedge_mapper/src/c8y/converter.rs:880]: Note also that this logic is not specific to Cumulocity and should not be here.

File: crates/core/thin_edge_json/src/event.rs
Comment (rina23q) [inline crates/core/thin_edge_json/src/event.rs:47]:
```suggestion
        JsonWriterError(#[from] json_writer::JsonWriterError),
```

Then, you can remove the line 33,
```rust
use json_writer::JsonWriterError;
```

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
This code must be refactored before merged. Notably `thin_edge_json` must be free of any constraints coming from Cumulocity.

An approach can be to add a source to each events.

```
/// In-memory representation of ThinEdge JSON event.
#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub struct ThinEdgeEvent {
    #[serde(rename = "type")]
    pub name: String,
    #[serde(flatten)]
    pub data: Option<ThinEdgeEventData>,
    pub source: Option<String>
}
```

So the `ThinEdgeEvent::try_from(event_topic, event_payload)` function can populate the source name from the event topic.

A new function of the c8y converter has to be added to convert the source name, if any, into an extra field of the even json.

Note also that there is no need of a `get_child_id_from_event_topic` function and that this decoding can be done by  the `ThinEdgeEvent::try_from(event_topic, event_payload)` function (with no third useless parameter).

File: crates/core/thin_edge_json/src/event.rs
Comment (rina23q) [inline crates/core/thin_edge_json/src/event.rs:58]: I feel this function's error handling is still weak.

It accepts these cases, but both must be errors:

- `topic_split.len() == 3` AND `may_be_child == Some("child-id")`
- `topic_split.len() == 4` AND `may_be_child == None`

Comment (rina23q) [reply crates/core/thin_edge_json/src/event.rs:58]: And I think these two cases are candidates of unit tests.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (rina23q) [inline crates/core/tedge_mapper/src/c8y/converter.rs:231]:
```suggestion
        // check if there is a child id in the topic, if not create the child before forwarding the event message
```

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (rina23q) [inline crates/core/tedge_mapper/src/c8y/converter.rs:232]: `child_id` can be `None`.
```suggestion
        let maybe_child_id = get_child_id_from_event_topic(&input.topic.name)?;
```

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (rina23q) [inline crates/core/tedge_mapper/src/c8y/converter.rs:892]: It's not Ok! If `topic.strip_prefix("tedge/events/")` returns `None`, that means the input topic name does not have `tedge/events/`, but this function is called! It's error.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (rina23q) [inline crates/core/tedge_mapper/src/c8y/converter.rs:882]: `v.len() >=2` means that this condition accepts `tedge/events/event_type/child_id/and_so_on`. It's an error.

**Review by rina23q — COMMENTED** [review]
Error handling inside functions need to be improved.
I haven't tested manually. I'll do it later.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (rina23q) [inline crates/core/tedge_mapper/src/c8y/converter.rs:878]: I think you wrote too verbose. It works the same.

```rust
pub fn get_child_id_from_event_topic(topic: &str) -> Result<Option<String>, ConversionError> {
    match topic.strip_prefix("tedge/events/").map(String::from) {
        Some(maybe_type_and_id) => {
            let v: Vec<&str> = maybe_type_and_id.split('/').collect();
            match v.len().cmp(&2) {
                Ordering::Equal if !v[1].is_empty() => Ok(Some(v[1].to_string())),
                _ => Err(ConversionError::InvalidChildId {
                    id: v[1].to_string(),
                }),
            }
        }
        None => Ok(None),
    }
}
```

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: Removed this function, refactored the code.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: This function does not exists any more.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs:232]: Removed this comment.

File: crates/core/thin_edge_json/src/event.rs
Comment (PradeepKiruvale) [reply crates/core/thin_edge_json/src/event.rs:57]: Now I derive the child id inside using the topic.

File: crates/core/thin_edge_json/src/event.rs
Comment (PradeepKiruvale) [reply crates/core/thin_edge_json/src/event.rs]: Moved

File: crates/core/thin_edge_json/src/event.rs
Comment (PradeepKiruvale) [reply crates/core/thin_edge_json/src/event.rs]: refactored.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: This function does not exists any more.

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
The code of `from_thin_edge_json_child_event` must be moved into `C8yCreateEvent::try_from`.
The function `try_convert_event` must be refactored to decouple the various aspects.

File: crates/core/c8y_translator/src/json.rs
Comment (PradeepKiruvale) [reply crates/core/c8y_translator/src/json.rs:71]: I did try let value = format!(r#"{{"externalId":"{source}","type":"c8y_serial"}}"#);

But was getting an extra field added for some reason the output looks like below, and the cumulocity throws an error if its in the below format.

[c8y/event/events/create] {"type":"login_event","time":"2021-01-01T05:30:45Z","text":"A user just logged in","externalSource":{"externalId":"child_p4","type":"c8y_serial"},**"type":"login_event"}**
The highlighted part was getting added at the end, but it should be before the `externalSource`.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: I created the function to create the `external source` if it does not exist. This already refactored a lot of code. So, felt its not necessary to create a method for creating the `smart rest message`.

File: crates/core/thin_edge_json/src/event.rs
Comment (PradeepKiruvale) [reply crates/core/thin_edge_json/src/event.rs]: updated the comment.

File: crates/core/tedge_mapper/Cargo.toml
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/Cargo.toml]: removed

File: crates/core/c8y_translator/src/json.rs
Comment (PradeepKiruvale) [reply crates/core/c8y_translator/src/json.rs]: Now replaced with JSON objects.

File: crates/core/c8y_translator/src/json.rs
Comment (PradeepKiruvale) [reply crates/core/c8y_translator/src/json.rs]: Yes, replaced with JSON object.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/converter.rs:268]: The code to send the event over HTTP doesn't depend on the child id - and must be outside the loop.

Furthermore this code is buggy. Since the child id has been added by `create_external_source_if_does_not_exist`, only the `else` case is used.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/converter.rs:235]:
```suggestion
        self.create_external_source_if_does_not_exist(&tedge_event, &mut vec);
```

**Review by didier-wenzek — CHANGES_REQUESTED** [review]

File: crates/core/c8y_translator/Cargo.toml
Comment (didier-wenzek) [inline crates/core/c8y_translator/Cargo.toml:11]: This seems to be unused

**Review by didier-wenzek — CHANGES_REQUESTED** [review]

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/converter.rs:254]: This is too fragile and obscure. Hence, I would look for another approach.

* The long comment is only required because the code is obscure.
* This is too fragile. Future extensions might break the assumption that `vec.len() == 1` implies `not registered`.
* Another source of fragility and obscurity: the check `message.payload_bytes().len() > self.size_threshold.0` is done in two places.

Why not, as we discussed, adding a `need_registration` boolean that makes things clearer?

```
let need_registration = self.create_external_source_if_does_not_exist(&tedge_event, &mut vec);

if message.payload_bytes().len() < self.size_threshold.0 {
   // The message can be sent via MQTT
  vec.push(message);
} else if !need_registration  {
  // The message must be sent over HTTP
  let _ = self.http_proxy.send_event(c8y_event).await?;
} else {
  // The message should be sent over HTTP but this cannot be done
 return Err(ConversionError::ChildDeviceNotRegistered {
                id: child_id.unwrap_or_else(|| "".into()),
            });
}
Ok(vec)
```

Furthermore, naming could help a lot to clarify the code.
* `messages` would be better than `vec`.
* `add_registration_messages` would be better than `create_external_source_if_does_not_exist`.
* `self.can_send_over_mqtt(&message)` would be better than `message.payload_bytes().len() < self.size_threshold.0`.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/converter.rs:230]:
```suggestion
        let c8y_event = C8yCreateEvent::try_from(tedge_event)?;
```

**Review by didier-wenzek — COMMENTED** [review]
This is far better. However there is still some room for improvements.

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: Addressed all of them

File: crates/core/tedge_mapper/src/c8y/converter.rs
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/converter.rs]: fixed

**Review by didier-wenzek — APPROVED** [review]
Approved

**Review by rina23q — CHANGES_REQUESTED** [review]
Unfortunately, I need to block merging because I found a bug.

Publish the topic: `tedge/events/empty/child_event`
Payload: `` (empty)

Then the mapper output is
```
400,empty,"empty",2022-07-21T15:47:29.445197198Z
```

The [specification](https://github.com/thin-edge/thin-edge.io/blob/cffd75655369381ae753593705f20b125cd4188c/docs/src/tutorials/send-events.md) says the Thin Edge JSON accepts the empty payload.

**Comment by PradeepKiruvale** [comment]
> Unfortunately, I need to block merging because I found a bug.
> 
> Publish the topic: `tedge/events/empty/child_id` Payload: (empty)
> 
> For example,
> 
> `tedge mqtt pub 'tedge/events/empty/child_id' ''`
> 
> Then the mapper output is
> 
> ```
> 400,empty,"empty",2022-07-21T15:47:29.445197198Z
> ```
> 
> The [specification](https://github.com/thin-edge/thin-edge.io/blob/cffd75655369381ae753593705f20b125cd4188c/docs/src/tutorials/send-events.md) says the Thin Edge JSON accepts the empty payload.

Maybe that is possible for the `parent` device. Not for the child device. Need to clarify with the specification writer!.

**Pull request merged**
